server/src/main/scala/org/http4s/server/middleware/CSRFMiddleware.scala

```diff
@@ -0,0 +1,196 @@
+    } yield joined + "-" + Base64.getEncoder.encodeToString(out)
+  }
+
+  /** Generate a new token **/
+  protected[middleware] def generateToken: Task[String] =
+    signToken(CSRFMiddleware.genTokenString)
+
+  /** Decode our CSRF token and extract the original token string to sign
+    */
+  protected[middleware] def extractRaw(token: String): OptionT[Task, String] =
+    token.split("-") match {
+      case Array(raw, nonce, signed) =>
+        val res = for {
+          instance <- Task.delay(Mac.getInstance(CSRFMiddleware.SigningAlgo))
+          _ <- Task.delay(instance.init(config.key))
+          out <- Task.delay(
+            instance.doFinal(
+              (raw + "-" + nonce).getBytes(StandardCharsets.UTF_8)))
+          r <- Task.point(
+            MessageDigest
+              .isEqual(out, Base64.getDecoder.decode(signed))
+              .option(raw))
+        } yield r
+        OptionT(res)
+      case _ =>
+        OptionT.none
+    }
+
+  /** Constructs a middleware that will check for the csrf token
+    * presence on both the proper cookie, and header values.
+    *
+    * If it is a valid token, it will then embed a new one,
+    * to effectively randomize the complete token while
+    * avoiding the generation of a new secure random Id, to guard
+    * against [BREACH](http://breachattack.com/)
+    *
+    */
+  def validate: Middleware[Request, MaybeResponse, Request, MaybeResponse] = {
+    service =>
+      Kleisli[Task, Request, MaybeResponse] { r =>
+        (for {
+          c1 <- CSRFMiddleware.cookieFromHeaders(r, config.cookieName)
+          c2 <- OptionT(
+            Task.point(r.headers.get(CaseInsensitiveString(config.headerName))))
+          raw1 <- extractRaw(c1.content)
+          raw2 <- extractRaw(c2.value)
+          response <- if (CSRFMiddleware.isEqual(raw1, raw2)) {
+            OptionT[Task, MaybeResponse](service(r).map(Some(_)))
+          } else {
+            OptionT.none[Task, MaybeResponse]
+          }
+          newToken <- OptionT[Task, String](signToken(raw1).map(Some(_)))
+        } yield
+          response.cata(
+            _.addCookie(Cookie(config.cookieName, newToken)),
+            Response(Status.NotFound)))
+          .getOrElse(Response(Status.Unauthorized))
+      }
+  }
+
+  /** Embed a token into a response **/
+  def embedNew(res: MaybeResponse): Task[Response] =
+    generateToken.map(
+      content =>
+        res.cata(_.addCookie(Cookie(config.cookieName, content)),
+          Response(Status.NotFound)))
+
+  /** Middleware to embed a csrf token into routes that do not have one. **/
+  def withNewToken: Middleware[Request, MaybeResponse, Request, MaybeResponse] =
+    _.andThen(Kleisli(embedNew))
+
+}
+
+object CSRFMiddleware {
+
+  /** An instance of SecureRandom to generate
+    * tokens, properly seeded:
+    * https://tersesystems.com/blog/2015/12/17/the-right-way-to-use-securerandom/
+    */
+  private lazy val CachedRandom = {
```

Comment:
Does this have to be lazy?

Reply:
SecureRandom will can block for the first use,while it gathers entropy, but I only want _one_ instance, since creation is expensive.

I don't feel strongly either way since objects themselves are lazy.